### PR TITLE
Inline most of the remaining functions in VtuStream into write_vtu_main().

### PR DESCRIPTION
**Pull request by bangerth**
Only the higher-order cell output is left now, and the use of `VtuStream` is confined to only a 4-line block.

Part of #14403.

/rebuild

### CODE REVIEW

File: source/base/data_out_base.cc [hunk 1]
```diff
@@ -6354,17 +6281,43 @@ namespace DataOutBase
                         const unsigned int d1 = 1;
                         const unsigned int d2 = n;
                         const unsigned int d3 = n * n;
+
+                        auto write_cell = [&flags, &out, &cells, d1, d2, d3](
+                                            const unsigned int start) {
```
Comment (drwells) [inline source/base/data_out_base.cc:6286]: ✔️

**Review by drwells — APPROVED** [review]

File: source/base/data_out_base.cc (hunk 1 above)
Comment (drwells) [reply source/base/data_out_base.cc:6286]: I like this style but we now get some unused variable warnings without zlib. Could you add

```cpp
(void)d1;
(void)d2;
(void)d3;
```
to the other case or some other equivalent fix?

**Comment by bangerth** [comment]
There are also warnings about lambda-capturing the variable
```
  const unsigned int d1 = 1;
```
Apparently, the compiler decides, from the initializer, that the variable is `constexpr` and that such variables are automatically captures. That's not a very useful warning.

I addressed all of this using a patch I had lined up separately in which I convert `DEAL_II_WITH_ZLIB` into a `constexpr` variable so that I can avoid the repeated preprocessor use. Let's see whether this works.

File: source/base/data_out_base.cc [hunk 2]
```diff
@@ -89,6 +89,13 @@ namespace
 
 namespace
 {
+#ifdef DEAL_II_WITH_ZLIB
+  constexpr bool deal_ii_with_zlib = true;
+#else
+  constexpr bool deal_ii_with_zlib = false;
+#endif
```
Comment (drwells) [inline source/base/data_out_base.cc:96]: once we require C++17 we can do this properly!

Comment (bangerth) [reply source/base/data_out_base.cc:96]: How would you do this with C++17? I'm translating a preprocessor symbol into a compiler variable, and I don't know a different way to do that.

There are places further down where we can use `if constexpr`, though.

**Comment by bangerth** [comment]
Ah, great. clang warns about variables that are automatically captured. But MSVC doesn't know about automatically capturing `constexpr` variables (a compiler error). One can't win here. I just got rid of these variables altogether.

File: source/base/data_out_base.cc (hunk 2 above)
Comment (drwells) [reply source/base/data_out_base.cc:96]: We wouldn't change this, but with C++17 we can replace a lot of preprocessor blocks (like the ones in this PR) with `if constexpr`.

Comment (bangerth) [reply source/base/data_out_base.cc:96]: Yes, true. In fact, we could change a lot of things in `config.h` where we currently `#define` something into `constexpr` something. I, for one, look forward to that!

**Pull request merged**